### ISSUE TITLE
Azure code coverage on private pipelines

### ISSUE THREAD

**Issue by cccs-cat001** [badges/shields#5083]
:question: **Question**

I'm new to shields, so forgive me if I'm super wrong. I'm trying to set up a code coverage shield for our azure pipeline, but I'm getting a badge `coverage: invalid`. I followed the directions and all. My only thought is there might be an issue with it being a private pipelines instance? Also I noticed that the URL that you call for the code coverage doesn't require the same parameters as you ask for? See [here to the rest of the file](https://github.com/badges/shields/blob/87fb882ef5baee341cdb79a682b4c637ce6281a4/services/azure-devops/azure-devops-coverage.service.js#L113). The endpoint microsoft says to use asks for a buildID, which references a specific build. But you are asking for essentially a pipelineID and branch. 

Any help would be greatly appreciated

**Comment by calebcartwright** [comment]
hi @cccs-cat001 :wave: 

tl;dr, you'll have to use a [self-hosted instance of the Shields server](https://github.com/badges/shields/blob/master/doc/self-hosting.md) if you want coverage and/or test result badges for private Azure DevOps projects; you cannot use the Shields.io service for private projects (this is generally true of all services Shields integrates with).

The main Shields.io service can only provide badges for public projects or cases where the upstream service/platform makes the requisite data available. The Shields.io service _can_ provide build and release status badges for private projects in Azure DevOps because Microsoft makes that data available anonymously (via the native status badges AzDO provides). 

> Also I noticed that the URL that you call for the code coverage doesn't require the same parameters as you ask for? See here to the rest of the file. The endpoint microsoft says to use asks for a buildID, which references a specific build. But you are asking for essentially a pipelineID and branch.

I'm not entirely sure what you are suggesting or asking there, so I'll instead give a recap of our code coverage badges in general and what's happening in the service class you linked.

The code coverage badges are used to dynamically visualize and display the current, most recent code coverage status, not a fixed/static percentage. In the case of Azure DevOps, this correlates to the code coverage from the latest build. 

As of the 5.1-preview API contract, Microsoft does not provide a single API that provides that data, which is why Shields is forced to make two API calls to Azure DevOps in sequence in order to dynamically get the code coverage for the _latest_ build. The first call, which is immediately above the line you linked to, is used to determine the buildid for the latest run of the specified pipeline:

https://github.com/badges/shields/blob/87fb882ef5baee341cdb79a682b4c637ce6281a4/services/azure-devops/azure-devops-coverage.service.js#L106-L112

and that build id is then used as a parameter in the second API call to get the corresponding code coverage:
https://github.com/badges/shields/blob/87fb882ef5baee341cdb79a682b4c637ce6281a4/services/azure-devops/azure-devops-coverage.service.js#L114-L120

The route parameters for the Shields badge are what Shields needs in order to make both of those API calls. 

This is what allows users to have a code coverage badge that's automatically updated.

**Comment by cccs-cat001** [comment]
Ah thank you! After I asked this question I did a deep dive through your code and I figured out how you get all the info you need to create the coverage badge. Thank you :)

**Comment by calebcartwright** [comment]
Happy to help! Going to close this issue accordingly
